Location reassignment add view for state users

mkangia opened the pull request:
https://dimagi-dev.atlassian.net/browse/ICDS-1339

##### SUMMARY
Adds a download only view for state users.
Updated feature flag to be both domain and user specific and then use that manage access to each view. Also thought of adding this to roles and hence manage via user permissions instead but its pretty custom to ICDS as of now, hence implemented it via feature flag only.

##### FEATURE FLAG
`LOCATION_REASSIGNMENT`

mkangia committed:
setup

allow bulk upload template to render form less template download only view

mkangia committed:
add download only view

mkangia committed:
add check for location access

mkangia committed:
mark view location_safe

orangejenny reviewed (changes requested):
Treating user flags and domain flags differently isn't intuitive (or documented - it would be hard for someone looking at this code to understand why it's set up how it is).

I also think that the location link in the reports menu should be limited only to people who can't access it anywhere else. For people who have access to locations, it's weird to see a locations link at the top of the reports menu.

What if you kept the flag user-based only, and then you displayed the link in the reports menu only `if not ProjectUsersTab(self._request)._is_viewable`?

mkangia commented on custom/icds/location_reassignment/views.py:
```diff
@@ -211,11 +214,10 @@ def _process_request_for_household_reassignment(self, parser, request, uploaded_
     location_id = request.GET.get('location_id')
 
-    if not location_id:
+    if not location_id or not user_can_access_location_id(domain, request.couch_user, location_id):
```
Ya. I assumed location search already handled that since its already used at so many places.
This is where i believe it handles that
https://github.com/dimagi/commcare-hq/blob/d2e1b01025435acbd33ed986c16ca9932b2bb68c/corehq/apps/reports/filters/controllers.py#L94

orangejenny replied:
sweet

mkangia commented on corehq/apps/hqwebapp/templates/hqwebapp/partials/bulk_upload.html:
```diff
@@ -42,21 +42,23 @@ <h2>
-      </p>
-      {% crispy bulk_upload_form %}
-    </li>
+    {% if bulk_upload_form %}
```
there are two steps, search location and then download so we are good here
offlined

mkangia commented:
@orangejenny we offlined that you would re-review this about the toggle is used, just following up on it.

mkangia commented:
@esoergel can you take a look at this and help us arrive at a conclusion,

in summary
1. the [current view](https://staging.commcarehq.org/a/myfirsttestproject/location_reassignment/) is behind a [user-based feature flag](https://staging.commcarehq.org/hq/flags/edit/location_reassignment/) which has all the power and comes under the locations tab
2. there is a need to have a view which web users/mobile users can login to ICDS and just download the file
3. these users have quite some minimum access available on HQ so the plan was to made it available under [Reports tab in side bar](https://staging.commcarehq.org/a/myfirsttestproject/location_reassignment_download/)

This PR
adds that new view and to make it available i have made the feature flag both domain and user specific and used that to make the download only view available if the feature is enabled for the domain and make the powerful complete view available in case the feature flag is enabled for the user.

@orangejenny and I offlined about it after her [review](https://github.com/dimagi/commcare-hq/pull/27357#pullrequestreview-404955124) not in favour of changing the FF and thought of making the download only view available at environment level instead and not change how the feature flag functions. But this makes this page available for all domains on an environment,, which is not nice too. So there is back and forth on it.

orangejenny commented:
@esoergel @mkangia 

> But this makes this page available for all domains on an environment

I might have misunderstood something when we offlined. Are you saying that if it's just a user flag, it'll show up for all domains on an environment for users who have it turned on? I think that's okay, although I'd also be fine with requiring **both** the user and domain flags to be on to see the link. (Although at that point maybe it should be a domain flag plus a user permission? I don't think it's that hard to add a new permission.) What I don't want is to support both the user and domain namespaces but have them enable different behavior (domain flag adds a link to the download in one place, user flag adds a link to the full feature in a different place).

esoergel commented:
The way I think of domain vs users tags is:

* Domain flags enable a feature for all users on the domain (with the appropriate role, if applicable).  These are vastly preferred for external usage, so Dimagi doesn't need to enable for each user, and so users don't see different things on the same project.
* User flags enable a feature for a single user on all domains.  This is useful for Dimagi users to be able to test out features before releasing, or to allow for advanced features we don't want to open up to partners.

In this case it looks like these are two different (though closely related) features, enabled depending on whether the flag is enabled for the user or for the domain.  I agree with Jenny that this violates expectations around how those work.

For example, the download-only view is linked in the menu if the flag is enabled by domain, but the view itself is also available if the flag is only enabled for the user.  That's a subset of the full feature, so it's not a big deal, but it'd similarly be easy to accidentally enable the reverse by using say `{% request|enabled %}` or `required_decorator`.

I'd prefer to split the two up, either as two feature flags or by enabling one for the ICDS environment - no strong preferences there.  You could also do something like enable the feature for the domain, but only grant full access to domain admins, or to users with full organization access.

Nice commit breakdown here, btw.

mkangia committed:
reveal links differently

Show download only link if the main view is not visible for user
Show main view only if feature flag set for both domain and user

mkangia commented:
thanks for weighing in @esoergel 

I offlined with @orangejenny and updated the [visibility of the links](https://github.com/dimagi/commcare-hq/pull/27357/commits/6e6b06b2a8d40d1b3af2a22021720ce2c02173a9). Didn't update the logic on the views because i though it was okay to not handle it there.
let me know how this looks now

mkangia commented:
> Nice commit breakdown here, btw.

Aha, thanks @esoergel

esoergel commented:
@mkangia what about the feature flag namespacing thing?  I do still think that's a violation of how feature flags are expected to work.

mkangia commented:
@esoergel you mean about creating two different feature flags instead of one?
If yes,
the logic that i have added would remain almost the same with the new feature flag being used for domain level check and the original feature flag for only user namespace test.

esoergel commented:
Yeah, two feature flags or one feature flag and a user role.  The logic seems fine, the part I object to is having different behavior depending on which feature flag namespace is enabled.

mkangia committed:
Merge branch 'master' into mk/location-reassignment-add-view

mkangia committed:
use different feature flag for download only view

also rename original feature flag

mkangia commented:
@esoergel updated. 
Thanks for the review

also pinging @orangejenny for a re-review.

mkangia commented:
@orangejenny @esoergel can you review this sometime today?
